### ISSUE TITLE
PDL::Primitive: `srand` fails to return the seed for perl >= 5.014

### ISSUE THREAD

**Issue by ewheelerinc** [PDLPorters/pdl#398]
Hello,

In PDL 2.080:

```sh
]$ perl  -MPDL -le 'print "PDL: $PDL::VERSION";'
PDL: 2.080
```

I was trying to get the seed from `srand()` to troubleshoot replaying the same Simplex optimization:

```sh
# perl -le 'print srand()'
3286181437
```
... but after PDL is loaded, srand() doesn't return the seed:

```sh
# perl -MPDL -le 'print srand()'
<nothing>
```

According to `perldoc -f srand`:

> When called with a parameter, "srand" uses that
> for the seed; otherwise it (semi-)randomly chooses a seed. In
> either case, starting with Perl 5.14, it returns the seed. To
> signal that your code will work only on Perls of a recent
> vintage:
>
>     use 5.014;  # so srand returns the seed

... so it looks like https://github.com/PDLPorters/pdl/blob/master/t/primitive.t#L89 needs to be updated to test for srand's return.

The `srand` function is implemented here: https://github.com/PDLPorters/pdl/blob/master/Basic/Primitive/primitive.pd#L2224

Beyond that I'm not sure how PDL's XS implementation plays out here, just that I was hoping to get a replay-capable implementation of `srand()` .  Not a big issue, but notworthy.

**Comment by ewheelerinc** [comment]
More info:

This may be a regression.  The issue above was opened with PDL 2.080 but on RHEL7 and PDL 2.032 it is working:

```sh
]$ perl -I /usr/lib64/perl5/vendor_perl/ -MPDL -le 'print "PDL: $PDL::VERSION";print srand()'
PDL: 2.032
3827158644
```

**Comment by zmughal** [comment]
Hi!

> ```shell
> # perl -le 'print srand()'
> 3286181437
> ```

This is `CORE::srand()`.

> ... but after PDL is loaded, srand() doesn't return the seed:
> 
> ```shell
> # perl -MPDL -le 'print srand()'
> <nothing>
> ```

This is `PDL::srand()` (imported into the `PDL` namespace when `PDL::Primitive` is loaded).

> According to `perldoc -f srand`:

You'll want to use `pdldoc` instead of `perldoc` to get the documentation for PDL's functions:

```shell
$ pdldoc srand
Module PDL::Primitive
  srand
      Signature: (a())

    Seed random-number generator with a 64-bit int. Will generate seed data
    for a number of threads equal to the return-value of "online_cpus" in
    PDL::Core.
```

There is sometimes some trickiness when mixing PDL code with other code since by default the `PDL` module imports a lot of things, so that is something to watch out for.

Of note, newer PDL versions use their own RNG because it is the same across platforms and supports use across threads (Perl's `rand()` uses various system provided RNGs depending on the configuration).

See [`Changes`](https://github.com/PDLPorters/pdl/blob/2.080/Changes#L211-L214)

https://github.com/PDLPorters/pdl/blob/0e8d4ee61a34c0505fa4adfb585846d7927635fb/Changes#L211-L214

The default way of seeding PDL is now `PDL::Core::seed()`, but this returns a different seed each time (mixing different sources of entropy from the system). So if you are using this in a test, just call `PDL::srand()` with any integer value (e.g., `PDL::srand(42)`).

I don't know if in the future we might change the underlying RNG to something better (once somebody develops a better RNG that fits our requirements of speed and enough entropy needed for scientific applications), but I think it should remain there for a while. I'm trying to think of a way of doing what R does which is let you set the RNG, so if we do add another RNG, I'll push to to offer an API to access the current RNG implementation as well.

Hope that helps!

P.S. Do you have a repo for <https://metacpan.org/dist/PDL-Opt-Simplex-Simple>?  I would love to add it to PDL's downstream testing (we test most of the CPAN modules that use PDL). We patch any upcoming regressions and it would be useful to know where to send the patches.

**Comment by ewheelerinc** [comment]
Ok, that makes sense.  We were running simplex a few times and sometimes we would get better results than others because it is deterministic but rand() isn't.  We would call srand() to store the seed in case someone wanted to use it and re-seed with the same value.  I'll switch it to PDL::srand(x) and store x for that purpose.

I think you can close this ticket and I'll open another about PDL::Opt::Simplex::Simple.

**Comment by ewheelerinc** [comment]
Oh, and FYI if anyone else wants the reference:

Bisect pinned this change down to 6be7747ab343beb8c034792cdac0cf81a4640bdf which was created to fix bug#350
